File: src/changefile/readChangeFiles.ts
import type { ChangeSet, ChangeInfo, ChangeInfoMultiple } from '../types/ChangeInfo';
import { getChangePath } from '../paths';
import fs from 'fs';
import path from 'path';
import type { BeachballOptions } from '../types/BeachballOptions';
import { getScopedPackages } from '../monorepo/getScopedPackages';
import { getChangesBetweenRefs } from 'workspace-tools';
import type { PackageInfos } from '../types/PackageInfo';
import { readJson } from '../object/readJson';

/**
 * Read change files, excluding any changes for packages that are:
 * - out of scope (as defined in `options.scope`)
 * - private
 * - nonexistent
 *
 * The changes will also be transformed if `options.transform.changeFiles` is provided.
 *
 * Changes from grouped change files will be flattened into individual entries in the returned array
 * (so it's possible that multiple entries will have the same filename).
 */
export function readChangeFiles(options: BeachballOptions, packageInfos: PackageInfos): ChangeSet {
  const { fromRef, command } = options;
  const scopedPackages = getScopedPackages(options, packageInfos);
  const changePath = getChangePath(options);

  if (!fs.existsSync(changePath)) {
    return [];
  }

  const allChangeFiles = fs.readdirSync(changePath);
  let filteredChangeFiles = allChangeFiles;

  if (fromRef) {
    const changeFilesSinceFromRef = getChangesBetweenRefs({
      fromRef,
      toRef: 'HEAD',
      options: [
        '--diff-filter=d', // excluding deleted files from the diff.
        '--relative', // results will include path relative to the cwd, i.e. only file names.
      ],
      pattern: '*.json',
      cwd: changePath,
    });

    filteredChangeFiles = allChangeFiles.filter(fileName => changeFilesSinceFromRef?.includes(fileName));
  }

  // sort the change files by modified time. Most recent modified file comes first.
  filteredChangeFiles.sort(
    (f1, f2) => getMtime({ changePath, changeFile: f2 }) - getMtime({ changePath, changeFile: f1 })
  );

  const changeSet: ChangeSet = [];

  // Read, transform, and filter the change files
  for (const changeFile of filteredChangeFiles) {
    const changeFilePath = path.join(changePath, changeFile);

    let changeInfo: ChangeInfo | ChangeInfoMultiple;
    try {
      changeInfo = readJson<ChangeInfo | ChangeInfoMultiple>(changeFilePath);
    } catch (e) {
      console.warn(`Error reading or parsing change file ${changeFilePath}: ${e}`);
      continue;
    }

    // Transform the change files, if the option is provided
    if (options.transform?.changeFiles) {
      try {
        changeInfo = options.transform?.changeFiles(changeInfo, changeFilePath, {
          command,
        });
      } catch (e) {
        console.warn(`Error transforming ${changeFilePath}: ${e}`);
        continue;
      }
    }

    const changes = (changeInfo as ChangeInfoMultiple).changes || [changeInfo as ChangeInfo];

    // Filter the changes from this file
    for (const change of changes) {
      // Log warnings about change entries for nonexistent and private packages.
      // (This may happen if a package is renamed or its private flag is changed.)
      const warningType = !packageInfos[change.packageName]
        ? 'nonexistent'
        : packageInfos[change.packageName].private
        ? 'private'
        : undefined;
      if (warningType) {
        const resolution = options.groupChanges ? 'remove the entry from this file' : 'delete this file';
        console.warn(
          `Change detected for ${warningType} package ${change.packageName}; ${resolution}: "${path.resolve(
            changePath,
            changeFile
          )}"`
        );
      }

      // Add the change to the final list if it's valid and in scope
      if (!warningType && scopedPackages.includes(change.packageName)) {
        changeSet.push({ changeFile, change });
      }
    }
  }

  return changeSet;
}

const mtimeCache: Record<string, number> = {};

/**
 * Get a file's modification time with caching. (Usually the caching won't matter, but it might
 * in large repos with many change files.)
 */
function getMtime(params: { changePath: string; changeFile: string }) {
  const cached = mtimeCache[params.changeFile];
  if (cached !== undefined) {
    return cached;
  }

  try {
    const mtime = fs.statSync(path.join(params.changePath, params.changeFile)).mtime.getTime();
    mtimeCache[params.changeFile] = mtime;
    return mtime;
  } catch (err) {
    mtimeCache[params.changeFile] = 0;
    return 0;
  }
}

File: rust/src/changefile/read_change_files.rs
use std::path::Path;

use crate::log_warn;
use crate::types::change_info::{ChangeFileInfo, ChangeInfoMultiple, ChangeSet, ChangeSetEntry};
use crate::types::options::BeachballOptions;
use crate::types::package_info::{PackageInfos, ScopedPackages};

/// Get the path to the change files directory.
pub fn get_change_path(options: &BeachballOptions) -> String {
    Path::new(&options.path)
        .join(&options.change_dir)
        .to_string_lossy()
        .to_string()
}

/// Read all change files from the change directory.
pub fn read_change_files(
    options: &BeachballOptions,
    package_infos: &PackageInfos,
    scoped_packages: &ScopedPackages,
) -> ChangeSet {
    let change_path = get_change_path(options);
    let change_dir = Path::new(&change_path);

    if !change_dir.exists() {
        return vec![];
    }

    let mut entries: Vec<(String, std::time::SystemTime)> = vec![];

    if let Ok(dir_entries) = std::fs::read_dir(change_dir) {
        for entry in dir_entries.flatten() {
            let path = entry.path();
            if path.extension().and_then(|e| e.to_str()) == Some("json") {
                let filename = path
                    .file_name()
                    .unwrap_or_default()
                    .to_string_lossy()
                    .to_string();
                let mtime = entry
                    .metadata()
                    .and_then(|m| m.modified())
                    .unwrap_or(std::time::SystemTime::UNIX_EPOCH);
                entries.push((filename, mtime));
            }
        }
    }

    // Sort by mtime (most recent first)
    entries.sort_by(|a, b| b.1.cmp(&a.1));

    let mut change_set = ChangeSet::new();

    for (filename, _) in entries {
        let file_path = change_dir.join(&filename);
        let contents = match std::fs::read_to_string(&file_path) {
            Ok(c) => c,
            Err(e) => {
                log_warn!("Error reading change file {filename}: {e}");
                continue;
            }
        };

        // Try parsing as grouped first, then single
        let changes: Vec<ChangeFileInfo> =
            if let Ok(multi) = serde_json::from_str::<ChangeInfoMultiple>(&contents) {
                multi.changes
            } else if let Ok(single) = serde_json::from_str::<ChangeFileInfo>(&contents) {
                vec![single]
            } else {
                log_warn!(
                    "{} does not appear to be a change file",
                    file_path.display()
                );
                continue;
            };

        for change in changes {
            // Warn about nonexistent/private packages
            let warning_type = if !package_infos.contains_key(&change.package_name) {
                Some("nonexistent")
            } else if package_infos[&change.package_name].private {
                Some("private")
            } else {
                None
            };

            if let Some(wt) = warning_type {
                let resolution = if options.group_changes {
                    "remove the entry from this file"
                } else {
                    "delete this file"
                };
                log_warn!(
                    "Change detected for {} package {}; {}: {}",
                    wt,
                    change.package_name,
                    resolution,
                    file_path.display()
                );
                continue;
            }

            if !scoped_packages.contains(&change.package_name) {
                continue;
            }

            change_set.push(ChangeSetEntry {
                change,
                change_file: filename.clone(),
            });
        }
    }

    change_set
}
